src/operations.rs

```diff
@@ -32,12 +32,24 @@ pub type SwapOperationUnchecked = SwapOperationBase<String>;
 pub type SwapOperation = SwapOperationBase<Addr>;
 
 impl SwapOperationUnchecked {
-    pub fn check(&self, api: &dyn Api) -> StdResult<SwapOperation> {
-        Ok(SwapOperation {
-            ask_asset_info: self.ask_asset_info.check(api)?,
-            offer_asset_info: self.offer_asset_info.check(api)?,
+    pub fn check(&self, deps: Deps) -> Result<SwapOperation, ContractError> {
+        let op = SwapOperation {
+            ask_asset_info: self.ask_asset_info.check(deps.api)?,
+            offer_asset_info: self.offer_asset_info.check(deps.api)?,
             pool: self.pool.clone(),
-        })
+        };
+        // validate pool assets
+        let pool_assets = op.pool.get_pool_liquidity(deps)?;
+        if !vec![op.offer_asset_info.clone(), op.ask_asset_info.clone()]
+            .iter()
+            .all(|a| pool_assets.find(a).is_some())
```

Comment:
Use `.contains()` instead?
```suggestion
        if !(pool_assets.contains(&op.offer_asset) && pool_asset.contains(&op.ask_asset))
```